src/Adapter/Order/Refund/OrderProductRemover.php

```diff
@@ -0,0 +1,265 @@
+        $order->total_products -= $shippingDiffTaxExcl;
+        $order->total_products_wt -= $productPriceTaxIncl;
+        $order->total_paid -= $productPriceTaxIncl + $shippingDiffTaxIncl;
+        $order->total_paid_tax_incl -= $productPriceTaxIncl + $shippingDiffTaxIncl;
+        $order->total_paid_tax_excl -= $productPriceTaxExcl + $shippingDiffTaxExcl;
+        $order->total_paid_real -= $productPriceTaxIncl + $shippingDiffTaxIncl;
+
+        $fields = [
+            'total_shipping',
+            'total_shipping_tax_excl',
+            'total_shipping_tax_incl',
+            'total_products',
+            'total_products_wt',
+            'total_paid',
+            'total_paid_tax_incl',
+            'total_paid_tax_excl',
+            'total_paid_real',
+        ];
+
+        /* Prevent from floating precision issues */
+        foreach ($fields as $field) {
+            if ($order->{$field} < 0) {
+                $order->{$field} = 0;
+            }
+        }
+
+        /* Prevent from floating precision issues */
+        foreach ($fields as $field) {
+            $order->{$field} = number_format(
+                $order->{$field},
+                Context::getContext()->getComputingPrecision(),
+                '.',
+                ''
+            );
+        }
+    }
+
+    /**
+     * @param Order $order
+     * @param OrderDetail $orderDetail
+     * @param int $quantity
+     * @param float $productPriceTaxIncl
+     * @param float $productPriceTaxExcl
+     * @param float $shippingDiffTaxIncl
+     * @param float $shippingDiffTaxExcl
+     *
+     * @throws DeleteProductFromOrderException
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function updateOrderDetail(
+        Order $order,
+        OrderDetail $orderDetail,
+        int $quantity,
+        float $productPriceTaxIncl,
+        float $productPriceTaxExcl,
+        float $shippingDiffTaxIncl,
+        float $shippingDiffTaxExcl
+    ) {
+        $orderDetail->product_quantity -= (int) $quantity;
+        if ($orderDetail->product_quantity == 0) {
+            if (!$orderDetail->delete()) {
+                throw new DeleteProductFromOrderException('Could not delete order detail');
+            }
+            if (count($order->getProductsDetail()) == 0) {
+                $history = new OrderHistory();
+                $history->id_order = (int) $order->id;
+                $history->changeIdOrderState(Configuration::get('PS_OS_CANCELED'), $order);
+                if (!$history->addWithemail()) {
+                    // email failure must not block order update process
+                    $this->logger->warning(
+                        $this->translator->trans(
+                            'Order history email could not be sent, test your email configuration in the Advanced Parameters > E-mail section of your back office.',
+                            [],
```

Comment:
ping @LouiseBonnard   

So I close the old PR with the new error message when the mail is not working.

In this PR, we do not block the feature is mail is not working, and we do not display an error message (a shop is not forced to have emails working). Instead we log an error message dedicated to the mail issue.

I was inspired by your old email error message, I only kept the part about checking the mail configuration.

Reply:
(this is a "technical" warning message, for shop administrator)

Reply:
Do you know what is the name of the email that could not have been sent in this case?

Reply:
It seems like the title of the mail is `[name of the store] Canceled"`

It is a mail that is sent when the status of an order has changed.